File: arena-jface/src/main/java/org/uqbar/lacar/ui/impl/jface/tables/JFaceTableBuilder.java
package org.uqbar.lacar.ui.impl.jface.tables;

import java.util.List;

import org.eclipse.jface.databinding.viewers.ViewersObservables;
import org.eclipse.jface.viewers.TableLayout;
import org.eclipse.jface.viewers.TableViewer;
import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Control;
import org.eclipse.swt.widgets.Table;
import org.uqbar.lacar.ui.impl.jface.JFaceContainer;
import org.uqbar.lacar.ui.impl.jface.JFaceControlBuilder;
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceBindingBuilder;
import org.uqbar.lacar.ui.model.BindingBuilder;
import org.uqbar.lacar.ui.model.ColumnBuilder;
import org.uqbar.lacar.ui.model.LabelProvider;
import org.uqbar.lacar.ui.model.TableBuilder;

import com.uqbar.commons.collections.CollectionFactory;

/**
 * Implementation of {@link TableBuilder} based on a JFace {@link TableViewer}.
 * 
 * @author npasserini
 * 
 * @param <R> The type of objects (models) shown in each row of the table.
 */
public class JFaceTableBuilder<R> extends JFaceControlBuilder<Table> implements TableBuilder<R> {
	private TableViewer viewer;
	private List<JFaceColumnBuilder<R>> columns = CollectionFactory.createList();
	private Class<R> itemType;

	// ********************************************************
	// ** Creation
	// ********************************************************

	public JFaceTableBuilder(JFaceContainer container, Class<R> itemType) {
		super(container);

		this.itemType = itemType;
		this.viewer = this.createTableViewer(this.getContainer().getJFaceComposite());
		this.initialize(this.getViewer().getTable());
	}

	protected org.eclipse.swt.graphics.Color getSWTColor(java.awt.Color color) {
		int blue = color.getBlue();
		int green = color.getGreen();
		int red = color.getRed();
		org.eclipse.swt.graphics.Color swtColor = new org.eclipse.swt.graphics.Color(getWidget().getDisplay(), red,
			green, blue);
		return swtColor;
	}

	private TableViewer createTableViewer(Composite jFaceComposite) {
		TableViewer viewer = new TableViewer(jFaceComposite, SWT.H_SCROLL | SWT.V_SCROLL | SWT.SINGLE | SWT.BORDER);
		viewer.getTable().setLinesVisible(true);
		viewer.getTable().setHeaderVisible(true);
		viewer.getTable().setLayout(new TableLayout());
		return viewer;
	}

	// ********************************************************
	// ** Description (columns)
	// ********************************************************

	@Override
	public ColumnBuilder<R> addColumn(LabelProvider<R> labelProvider) {
		JFaceColumnBuilder<R> column = new JFaceColumnBuilder<R>(this, labelProvider);
		this.columns.add(column);
		return column;
	};

	// ********************************************************
	// ** Binding
	// ********************************************************

	@Override
	public BindingBuilder observeValue() {
		return new JFaceBindingBuilder(this, ViewersObservables.observeSingleSelection(this.getViewer()));
		// return new JFaceBindingBuilder(this, new TableObservableValue(this.getWidget(), this.itemType));
	};

	@Override
	public BindingBuilder observeContents() {
		return new JFaceTableItemsBindingBuilder<R>(this);
	}

	// ********************************************************
	// ** Compilation
	// ********************************************************

	@Override
	public void pack() {
		this.getViewer().getTable().setLayout(new JFaceTableLayoutBuilder(this).createLayout());

		for (JFaceColumnBuilder<R> column : this.columns) {
			column.pack();
		}
		super.pack();
	}

	// ********************************************************
	// ** Internal accessors
	// ********************************************************

	@Override
	protected Control getControlLayout() {
		return this.getViewer().getControl();
	}

	public TableViewer getJFaceTableViewer() {
		return this.getViewer();
	}

	public List<JFaceColumnBuilder<R>> getColumns() {
		return this.columns;
	}

	public Class<R> getItemType() {
		return this.itemType;
	}

	public TableViewer getViewer() {
		return viewer;
	}
}

File: arena-jface/src/main/scala/org/uqbar/lacar/ui/impl/jface/builder/tables/JFaceTableBuilder.scala
package org.uqbar.lacar.ui.impl.jface.builder.tables

import java.util.ArrayList
import scala.collection.JavaConversions.asScalaBuffer
import org.eclipse.jface.databinding.viewers.ViewersObservables
import org.eclipse.jface.viewers.TableLayout
import org.eclipse.jface.viewers.TableViewer
import org.eclipse.swt.SWT
import org.eclipse.swt.layout.GridData
import org.eclipse.swt.layout.GridLayout
import org.eclipse.swt.layout.RowData
import org.eclipse.swt.widgets.Composite
import org.eclipse.swt.widgets.Table
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceBindingBuilder
import org.uqbar.lacar.ui.impl.jface.builder.JFaceControlBuilder
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceContainer
import org.uqbar.lacar.ui.impl.jface.tables.JFaceTableItemsBindingBuilder
import org.uqbar.lacar.ui.impl.jface.tables.JFaceTableLayoutBuilder
import org.uqbar.lacar.ui.model.LabelProvider
import org.uqbar.lacar.ui.model.TableBuilder
import org.eclipse.swt.graphics.Color
import org.eclipse.swt.layout.FillLayout
import org.eclipse.swt.layout.RowLayout

class JFaceTableBuilder[R](container: JFaceContainer, var itemType: Class[R])
  extends JFaceControlBuilder[Table](container)
  with TableBuilder[R] {

  var columns = new ArrayList[JFaceColumnBuilder[R]]
  var viewer = this.createTableViewer(container.getJFaceComposite)
  
  initialize(viewer.getTable)
  setNumberVisibleRows(3)

  def createTableViewer(jFaceComposite: Composite) = {
    val viewer = new TableViewer(new Composite(jFaceComposite, 0), SWT.H_SCROLL | SWT.V_SCROLL | SWT.SINGLE | SWT.BORDER)
    viewer.getTable.setLinesVisible(true)
    viewer.getTable.setHeaderVisible(true)
    viewer.getTable.setLayout(new TableLayout())
    viewer
  }

  override def addColumn(labelProvider: java.util.List[LabelProvider[R]]) = {
    val column = new JFaceColumnBuilder[R](this, labelProvider)
    columns.add(column)
    column
  }

  override def observeValue = new JFaceBindingBuilder(this, ViewersObservables.observeSingleSelection(viewer))
  override def observeContents = new JFaceTableItemsBindingBuilder[R](this)

  override def getControlLayout = viewer getControl

  def setNumberVisibleRows(numberVisibleRows: Int) {
    val layout = new FillLayout
    
    viewer.getControl().getParent().setBackground(new Color(viewer.getControl().getDisplay(),0,0,0))
    
    viewer.getControl().getParent().setLayout(layout)
    viewer.getTable.setLayout(new JFaceTableLayoutBuilder(this).createLayout)    
    columns foreach (_ pack)
    
    
    val rd = new RowData()
    rd.height = viewer.getTable().getItemHeight() * (numberVisibleRows + 1) 
    viewer.getControl().getParent().setLayoutData(rd)
    
    var p = viewer.getControl().getParent()
    while(p != null){
      p.layout()
      p = p.getParent()
    }
    
  }

  override def pack() = {
    viewer.getTable.setLayout(new JFaceTableLayoutBuilder(this).createLayout)    
    columns foreach (_ pack)
    
    super.pack()
  }

}